File: algo-calculate-mode/python/calculate_mode.py
def calculate_mode(array):
    mode = []
    max_element_count = 0
    element_counts = {}
    for element in array:
        element_counts[element] = element_counts.get(element, 0) + 1
    for elem_count in element_counts:
        max_element_count = max(element_counts[elem_count], max_element_count)
    for elem_count in element_counts:
        if element_counts[elem_count] == max_element_count:
            mode.append(elem_count)

    return mode


print(calculate_mode([1, 2, 3, 3]))

File: algo-calculate-mode/js/calculateMode.js
const calculateMode = (list) => {
  const mode = [];
  let max_count = 0;
  const element_counts = {};
  list.forEach(
    (elem) => (element_counts[elem] = element_counts[elem] + 1 || 1)
  );
  for (let elem in element_counts) {
    if (element_counts[elem] > max_count) {
      max_count = element_counts[elem];
    }
  }
  for (let elem in element_counts) {
    if (element_counts[elem] === max_count) {
      mode.push(elem);
    }
  }
  return mode.map((elem) => {
    if (isNaN(Number(elem))) {
      return elem;
    } else {
      return Number(elem);
    }
  });
};

module.exports = { calculateMode };
